<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.8.0 -->
<title>Implementation | sgcount</title>
<meta name="generator" content="Jekyll v3.9.2" />
<meta property="og:title" content="Implementation" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="A fast and flexible sgRNA counter for CRISPR screens." />
<meta property="og:description" content="A fast and flexible sgRNA counter for CRISPR screens." />
<link rel="canonical" href="http://localhost:4000/sgcount.github.io/implementation/" />
<meta property="og:url" content="http://localhost:4000/sgcount.github.io/implementation/" />
<meta property="og:site_name" content="sgcount" />
<meta property="og:type" content="website" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="Implementation" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"WebPage","description":"A fast and flexible sgRNA counter for CRISPR screens.","headline":"Implementation","url":"http://localhost:4000/sgcount.github.io/implementation/"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/sgcount.github.io/assets/main.css"><link type="application/atom+xml" rel="alternate" href="http://localhost:4000/sgcount.github.io/feed.xml" title="sgcount" /></head>
<body><header class="site-header" role="banner">

  <div class="wrapper"><a class="site-title" rel="author" href="/sgcount.github.io/">sgcount</a><nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger"><a class="page-link" href="/sgcount.github.io/about/">About</a><a class="page-link" href="/sgcount.github.io/implementation/">Implementation</a><a class="page-link" href="/sgcount.github.io/install/">Install</a></div>
      </nav></div>
</header>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <article class="post">

  <header class="post-header">
    <h1 class="post-title">Implementation</h1>
  </header>

  <div class="post-content">
    <hr />
<h2 id="implementation">Implementation</h2>
<p>Ultimately this is written to be as fast as possible by using sequence hashing.</p>

<p>The library is hashed upon entry, sequences are trimmed to the positions of interest, then checked against the hashmap and counted.</p>

<h3 id="reading-in-the-library">Reading in the Library</h3>
<p>The library is read-in as a <code class="language-plaintext highlighter-rouge">fasta</code> formatted file into a <code class="language-plaintext highlighter-rouge">HashMap</code>.</p>

<p>It is checked for uniqueness and requires each sequence to have an equal length.</p>

<p>If duplicates sequences are found <code class="language-plaintext highlighter-rouge">sgcount</code> will quit.</p>

<h3 id="mismatch-library">Mismatch Library</h3>
<p>To allow for one-offs in the reference library I create an unambiguous one-off hashmap by using the sequences within the library.</p>

<p>In essence, each sequence in the library has every valid 1-hamming-distance sequence generated.</p>

<p>Then any of those generated sequences which could map back to more than one reference sequence is removed.</p>

<p>This front-loads the search by pre-generating all the possible one-offs, and still allows O(1) search when iterating over the sequences.</p>

<h3 id="variable-region-positioning-and-reverse-complement-identification">Variable-Region Positioning and Reverse-Complement Identification</h3>
<p>I used an <a href="https://en.wikipedia.org/wiki/Entropy_(information_theory)">entropy</a>-based approach to determine where the variable region would be in the sequencing dataset.</p>

<p>Because the sequences are constructed with a constant region it means that the nucleotide entropy (i.e. the randomness in observing a nucleotide <code class="language-plaintext highlighter-rouge">A</code>, <code class="language-plaintext highlighter-rouge">C</code>, <code class="language-plaintext highlighter-rouge">T</code>, <code class="language-plaintext highlighter-rouge">G</code>) in that region will be considerably lower than in the variable region.</p>

<p>As a result, if you were to plot the entropy in each position across the sequence you will see near-zero entropy everywhere except for the variable region.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[ 1e-3 1e-3 2e-3 1e-3 0.5 0.7 0.8 0.7 1e-4 1e-3 1e-3 2e-3 ]
                     ^               ^
                     |               |
             variable_start    variable_end
</code></pre></div></div>

<p>To determine where this region starts and ends, I first calculate the positional entropy in the reference library, then sample a few thousand sequences from the sample libraries, and then determine which positions minimize a minimum-squared-error between the two.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[ 0.5 0.7 0.8 0.7 ]                       &lt;&lt; Reference Entropy
[ 1e-3 2e-3 0.5 0.7 0.8 0.7 1e-4 1e-3 ]   &lt;&lt;  Observed Entropy
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[ 1e-3 2e-3 0.5 0.7 0.8 0.7 1e-4 1e-3 ]
[ 0.5  0.7  0.8 0.7 ]
     [ 0.5  0.7 0.8 0.7 ]
          [ 0.5 0.7 0.8 0.7 ]           &lt;&lt; Minimal MSE @ 3rd basepair
              [ 0.5 0.7 0.8 0.7 ]
                  [ 0.5 0.7 0.8  0.7 ]
</code></pre></div></div>

<p>This strategy easily lets you determine the reverse complement as the positional entropy vector can just be flipped and searched in the same manner.</p>

<h3 id="single-position-misplacement">Single-Position Misplacement</h3>
<p>To allow for indels before the variable region I’ve implemented a single-position misplacement allowment.</p>

<p>Visually this looks like this:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ACTGGACAA|VARIABLE|
ACTGGACAA|VARIABLE|
ACTGGACAA|VARIABLE|
ACTGGACA|VARIABLE|     &lt;&lt; Single deletion
ACTGGACAAA|VARIABLE|   &lt;&lt; Single insertion
ACTGGAC|VARIABLE|      &lt;&lt; Double deletion (not supported)
ACTGGACAAAA|VARIABLE|  &lt;&lt; Double insertion (not supported)

</code></pre></div></div>

<p>Since the reads are trimmed to the expected variable length before searching the hashmap, these last two examples would’ve been missed even though their variable sequence occured without any mismatches.</p>

<p>To capture these sequences the reads are first tried at the expected variable position, but if there are no matches they are retried with a <code class="language-plaintext highlighter-rouge">-1</code> variable position, and again with a <code class="language-plaintext highlighter-rouge">+1</code> variable position.</p>

<p>If both of these fail then it is considered a true failure and will not be captured by <code class="language-plaintext highlighter-rouge">sgcount</code>.</p>

<p>This could be repeated <em>ad infinitum</em> but I think it starts to lose biological credibility as the indel size increases.</p>

<p>In my experience this has recovered an extra 3-5% of sequences, which is not insignificant, and could make a very large difference if those sgRNAs were low abundance to begin with.</p>

  </div>

</article>

      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/sgcount.github.io/"></data>

  <div class="wrapper">

    <h2 class="footer-heading">sgcount</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li class="p-name">sgcount</li><li><a class="u-email" href="mailto:noam.teyssier@ucsf.edu">noam.teyssier@ucsf.edu</a></li></ul>
      </div>

      <div class="footer-col footer-col-2"><ul class="social-media-list"><li><a href="https://github.com/noamteyssier"><svg class="svg-icon"><use xlink:href="/sgcount.github.io/assets/minima-social-icons.svg#github"></use></svg> <span class="username">noamteyssier</span></a></li><li><a href="https://www.twitter.com/noamteyssier"><svg class="svg-icon"><use xlink:href="/sgcount.github.io/assets/minima-social-icons.svg#twitter"></use></svg> <span class="username">noamteyssier</span></a></li></ul>
</div>

      <div class="footer-col footer-col-3">
        <p>A fast and flexible sgRNA counter for CRISPR screens. </p>
      </div>
    </div>

  </div>

</footer>
</body>

</html>
