<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.8.0 -->
<title>About | sgcount</title>
<meta name="generator" content="Jekyll v3.9.2" />
<meta property="og:title" content="About" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="A fast and flexible sgRNA counter for CRISPR screens." />
<meta property="og:description" content="A fast and flexible sgRNA counter for CRISPR screens." />
<link rel="canonical" href="http://localhost:4000/sgcount.github.io/about/" />
<meta property="og:url" content="http://localhost:4000/sgcount.github.io/about/" />
<meta property="og:site_name" content="sgcount" />
<meta property="og:type" content="website" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="About" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"WebSite","description":"A fast and flexible sgRNA counter for CRISPR screens.","headline":"About","name":"sgcount","url":"http://localhost:4000/sgcount.github.io/about/"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/sgcount.github.io/assets/main.css"><link type="application/atom+xml" rel="alternate" href="http://localhost:4000/sgcount.github.io/feed.xml" title="sgcount" /></head>
<body><header class="site-header" role="banner">

  <div class="wrapper"><a class="site-title" rel="author" href="/sgcount.github.io/">sgcount</a><nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger"><a class="page-link" href="/sgcount.github.io/about/">About</a><a class="page-link" href="/sgcount.github.io/implementation/">Implementation</a><a class="page-link" href="/sgcount.github.io/install/">Install</a></div>
      </nav></div>
</header>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <article class="post">

  <header class="post-header">
    <h1 class="post-title">About</h1>
  </header>

  <div class="post-content">
    <hr />
<h2 id="background">Background</h2>
<p>CRISPR screening is a powerful platform in functional genomics to determine which genes affect a quantitative phenotype.
In traditional 1D screens this phenotype takes the form of survival, fluorescence, phagocytosis etc.
Once the cells have been processed and sequenced you will have multiple libraries and sequencing files associated with each of them.</p>

<p>If you are performing a CRISPRi/a v2 screen your sequences will be constructed with the following format:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[Constant-Adapter] - [Variable-Region] - [Constant-Adapter]
</code></pre></div></div>

<p>Depending on how you prepared your libraries however, your sequences will have a variable sized constant region, and may be reverse complemented with respect to your library.</p>

<hr />

<h2 id="analysis">Analysis</h2>
<p>The ultimate goal of this tool is to count the number of each sgRNA within each sample.</p>

<h3 id="steps">Steps</h3>
<ol>
  <li>
    <p>Variable region placement is identified (this is done on a sample-by-sample basis in case any samples were prepared differently).</p>
  </li>
  <li>Library sequences are read in and validated for uniqueness and equivalent size.
    <ul>
      <li>This program <strong>will not run</strong> if there are <strong>duplicate sequences</strong> (solutions in <a href="/usage/">usage</a>).</li>
      <li>This program <strong>will not run</strong> if the <strong>library sequences are different lengths</strong>.</li>
      <li>If single mismatches are allowed the unambiguous mismatch library will be generated.</li>
    </ul>
  </li>
  <li>
    <p>Sample variable-region positions are calculated and reverse-complement status is determined.</p>
  </li>
  <li>Sequences in each sample are left-trimmed to remove the adapter and right-trimmed to the sequence length of the library.
    <ul>
      <li>If reverse complement is required the reads are reverse complemented.</li>
      <li>If single misplace is allowed, positional mismatches are retried with ±1 basepair in the variable-region position.</li>
    </ul>
  </li>
  <li>Count tables are written.</li>
</ol>

<hr />

<h2 id="features">Features</h2>
<ul>
  <li>Automatic adapater location identification</li>
  <li>Automatic reverse complement identification</li>
  <li>Single mismatch allowance (one-off in sgRNA Sequence)</li>
  <li>Single misplace allowance (variable region position shift)</li>
  <li>Multiple samples at once</li>
  <li>Multiple cores</li>
</ul>

<hr />

<h2 id="requirements">Requirements</h2>
<ul>
  <li>sequencing data (<code class="language-plaintext highlighter-rouge">*.fastq</code>/<code class="language-plaintext highlighter-rouge">*.fastq.gz</code>)</li>
  <li>reference library (<code class="language-plaintext highlighter-rouge">*.fasta</code>/<code class="language-plaintext highlighter-rouge">*.fasta.gz</code>)</li>
  <li>sgRNA-Gene Mapping (<code class="language-plaintext highlighter-rouge">*.txt</code>) [Optional]</li>
</ul>

<hr />

<h2 id="implementation">Implementation</h2>
<p>If you’re interested to see how this works, or want to contribute, feel free to check out the <a href="/implementation/">implementation details</a>.</p>

  </div>

</article>

      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/sgcount.github.io/"></data>

  <div class="wrapper">

    <h2 class="footer-heading">sgcount</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li class="p-name">sgcount</li><li><a class="u-email" href="mailto:noam.teyssier@ucsf.edu">noam.teyssier@ucsf.edu</a></li></ul>
      </div>

      <div class="footer-col footer-col-2"><ul class="social-media-list"><li><a href="https://github.com/noamteyssier"><svg class="svg-icon"><use xlink:href="/sgcount.github.io/assets/minima-social-icons.svg#github"></use></svg> <span class="username">noamteyssier</span></a></li><li><a href="https://www.twitter.com/noamteyssier"><svg class="svg-icon"><use xlink:href="/sgcount.github.io/assets/minima-social-icons.svg#twitter"></use></svg> <span class="username">noamteyssier</span></a></li></ul>
</div>

      <div class="footer-col footer-col-3">
        <p>A fast and flexible sgRNA counter for CRISPR screens. </p>
      </div>
    </div>

  </div>

</footer>
</body>

</html>
